python/nwm_client_new/src/hydrotools/nwm_client_new/NWMClient.py

```diff
@@ -60,20 +63,53 @@ class NWMClientDefaults:
         "streamflow": "value"
     })
     SSL_CONTEXT: ssl.SSLContext = ssl.create_default_context()
-    def __post_init__(self):
-        # Gather routelink files
-        rl_filepath = Path(__file__).parent / "data/routelink_files"
-        rl_files = rl_filepath.glob("*.csv")
-
-        # Generate crosswalk
-        dfs = []
-        for rl_file in rl_files:
-            dfs.append(pd.read_csv(
-            rl_file,
-            dtype={"nwm_feature_id": int, "usgs_site_code": str},
-            comment='#'
-        ).set_index('nwm_feature_id')[['usgs_site_code']])
-        self.CROSSWALK = pd.concat(dfs)
+    ROUTELINK_URL: str = "https://www.hydroshare.org/resource/d154f19f762c4ee9b74be55f504325d3/data/contents/RouteLink.h5"
+
+    def get_routelink(
+        self, 
+        url: str,
+        ssl_context: ssl.SSLContext
+        ) -> dd.DataFrame:
+        """Retrieve NWM RouteLink data from URL and return a 
+        dask.dataframe.DataFrame.
+        
+        Parameters
+        ----------
+        url: str
+            URL path to HDF5 RouteLink file.
+        ssl_context: ssl.SSLContext
+            SSL Context instance.
+            
+        Returns
+        -------
+        df: dask.dataframe.DataFrame
+            DataFrame containing associated location metadata.
+        """
+        with TemporaryDirectory() as td:
+            # Setup downloader
+            downloader = FileDownloader(
+                output_directory=td,
+                create_directory=False,
+                ssl_context=ssl_context
+                )
+
+            # Download files
+            downloader.get([(url, "RouteLink.h5")])
+            return dd.from_pandas(pd.read_hdf(Path(td)/"RouteLink.h5"), 
+                npartitions=1)
+    
+    @property
+    def CROSSWALK(self) -> pd.DataFrame():
```

Comment:
Is the call to `pd.DataFrame()` desired? Ive never seen this style of type hint, so I am more curious than anything.

Reply:
Given that we might want to support versions of the crosswalk in the future, it might be a good idea to support that. So, maybe something like move the innards the `CROSSWALK` property to a `get_crosswalk` method that takes in a version number (enum of versions we support) and returns that crosswalk. It would be great to then tag the dataframe with metadata (maybe `df.attrs["version"] = "2.1"`. `attrs`' is still not "stable", im just spitballing). Then the `CROSSWALK` property just returns the latest crosswalk. thoughts?

Reply:
That type hint is an error. I'm kind of surprised that works.

I agree with versioning in principle, but I'm not sure how achievable it is in practice. You already mentioned `attrs` still being unstable, but we also have a limitation on how NWM treats "versions." For example, version 2.2 of the NWM is set to release soon, but there will be no version 2.2 of auxiliary files. NWM 2.2 will be backwards compatible with all auxiliary files including RouteLinks.

There's an additional complication in that the client tools is compatible with the 2.0 NWM data on Google Cloud, even though it uses the NWM 2.1 RouteLink information. The distinction between NWM versions and NWM RouteLink versions is muddy.

I've added this topic as an agenda item for our next meeting. This might be a question we can pose to the Next Gen NWM team.

Reply:
Right, model versioning does not necessarily equal to parameter versioning. Ill close this and we can discuss this in the upcoming meeting. I dont think we want to become the source of "truth" for the mapping of nwm version's to parameter version's, hopefully we can find another source where that is documented and point people there and/or mirror that "truth".